src/3rd_party_repos.c

```diff
@@ -0,0 +1,132 @@
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "3rd_party_repos.h"
+#include "swupd.h"
+
+#include <errno.h>
+
+#ifdef THIRDPARTY
+
+static char *get_repo_config_path(void)
+{
+	return str_or_die("%s/%s/%s", globals.state_dir, "3rd_party", "repo.ini");
+}
+
+/**
+ * @brief This function is called by the repo ini parse.
+ *
+ * @param section A string containing section of the file
+ * @param key A string containing key
+ * @param value A string containing value
+ *
+ * @returns True for every valid line which could be extracted, false otherwise
+ */
+static bool parse_key_values(char *section, char *key, char *value, void *data)
+{
+	struct repo *repo;
+	struct list **repos = data;
+
+	//TODO: Extend to support other fields on the config file too.
+	//TODO: Don't load duplicated entries in the list
+	if (strcasecmp(key, "url") == 0) {
+		repo = calloc(1, sizeof(struct repo));
+		ON_NULL_ABORT(repo);
+		repo->name = strdup_or_die(section);
+		repo->url = strdup_or_die(value);
+		*repos = list_append_data(*repos, repo);
+	}
+
+	return true;
+}
+
+struct list *third_party_get_repos(void)
+{
+	char *repo_config_file_path = get_repo_config_path();
+	struct list *repos = NULL;
+
+	if (!config_file_parse(repo_config_file_path, parse_key_values, &repos)) {
+		list_free_list_and_data(repos, repo_free_data);
+		repos = NULL;
+	}
+
+	repos = list_head(repos);
+	free_string(&repo_config_file_path);
+
+	return repos;
+}
+
+static int repo_name_cmp(const void *repo, const void *name)
+{
+	return strcmp(((struct repo *)repo)->name, name);
+}
+
+void repo_free(struct repo *repo)
+{
+	free(repo->name);
+	free(repo->url);
+}
+
+void repo_free_data(void *data)
+{
+	struct repo *repo = (struct repo *)data;
+	repo_free(repo);
+}
+
+int third_party_add_repo(char *repo_name, char *repo_url)
+{
+	int ret = 0;
+	struct list *repos;
+	char *repo_config_file_path = NULL;
+	FILE *fp = NULL;
+
+	repos = third_party_get_repos();
+	if (list_search(repos, repo_name, repo_name_cmp)) {
+		error("The repo: %s already exists\n", repo_name);
+		ret = -ENOENT;
+		goto exit;
+	}
+
+	// If we are here, we are cleared to write to file
+	repo_config_file_path = get_repo_config_path();
+	fp = fopen(repo_config_file_path, "a");
+	if (!fp) {
+		ret = -errno;
+	}
+
+	// write name
+	ret = config_write_section(fp, repo_name);
+	if (ret) {
+		goto exit;
+	}
+
+	// write url
+	ret = config_write_config(fp, "url", repo_url);
```

Comment:
I think we should validate repo_url for two things here before writing it to the config file:
- we should validate the user has specified a protocol (http, https or file), that way we minimize the possible user error of `swupd 4rd-party add my-repo myrepo.com`
- we should warn the user about the risks of using an insecure protocol if he/she decides to use http

Reply:
> we should validate the user has specified a protocol (http, https or file), that way we minimize the possible user error of swupd 4rd-party add my-repo myrepo.com

We should be consistent and we don't validate that in mirror --set, so we sholdn't validate that in 3rd-party repo set. I don't like validating that too. We can trust curl to know how to download from the correct place and print warnings accordingly. One possibility would be to check if the connection is valid here, but I personally prefer to validate that only when needed. (kind of like git remote). If we let curl take care of the protocol all new protocols supported by curl will be automatically supported by swupd for free.

> we should warn the user about the risks of using an insecure protocol if he/she decides to use http

That's true. We need that. I'll add.